src/nvim/os/env.c

```diff
@@ -169,6 +178,64 @@ void os_get_hostname(char *hostname, size_t size)
 #endif
 }
 
+/// @return -1 if error encountered, 1 if capslock active otherwise 0 if not.
+int capslock_status(void ) {
+
+#if defined(__linux__)
+
+  FILE *fp = NULL;
+  char res[1024];
+  fp = popen("xset -q | grep Caps", "r");
```

Comment:
This also assumes that xset is installed (I have it in a separate package with only two dependents (though one of which is pretty common xdg-utils)) and in $PATH, that shell is POSIX compatible to some extent and that grep is there and in $PATH as well. And, BTW, `popen` is not C99, it is POSIX; and there is libuv for that. I do not like shelling out in C code in any case, but there is no better alternative to xset it code must be limited to calling just `xset` and nothing else, not even a shell. In any case there is no way `grep` will beat even Neovim regexes as they are already loaded, and it does not look like regexes are what is actually needed.

Reply:
Yes, missed it, and also I didn't like `xset` because this is some external application, I was thinking whatif they change their format later and also the installation of xset was also be required, this was just a prototype, for just testing if it was working on my pc or not, so it worked, I will edit it today after the GSoC proposal is finalized

Reply:
I don't think `xsel` should be hardcoded, it should work more like say clipboard provider. If there is a similar utility for wayland it could then be configured. Something like `os_system()` would be preferable over `popen()`.

Reply:
Yes, I have to work on it and find it, I will try my best

Reply:
I came to know that its not possible to find out the status of capslock in console mode, I also asked on linux IRC, they said same, for reading the capslock status, I need to hook into the hid interface which may require some special permissions, so initially I should make this function for linux XWindows interface only, moreover I want to say, and encourage to try, try to open console  in linux by `ctrl+alt+f1` login there, then add command `xset -q` this will give error, more over when you press the capslock key, the led for the capslock will not be lighted, so there may be some issues with capslock.
RFC for this discussion